File: problems/python/linked_list/mergeKSortedLists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
import heapq

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        heap = []
        res = None
        temp = None

        for index, node in enumerate(lists):
            if not node == None:
                heapq.heappush(heap,(node.val, index))
        
        while len(heap) > 0:
            cur = heapq.heappop(heap)
            curNode = lists[cur[1]]
            lists[cur[1]] = curNode.next
            curNode.next = None

            if temp == None:
                temp = curNode
                res = temp
            else:
                temp.next = curNode
                temp = temp.next

            if lists[cur[1]] != None:
                cur = (lists[cur[1]].val,cur[1])
                heapq.heappush(heap, cur)
        
        return res
        

File: java/linked_list/merge_k_sorted_lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<int[]> heap = new PriorityQueue<>( (a,b) -> Integer.compare(a[0],b[0]) );
	    ListNode sol = null;
        ListNode res = null;

        for(int i=0; i<lists.length; i++) {
	    	if (lists[i] != null) {
	    		int[] cur = new int[2];
	    		cur[0] = lists[i].val;
	    		cur[1] = i;
	    		heap.add(cur); // O(logn)
            }
        }

        while(heap.size() != 0) {
        	int[] cur = heap.poll();
        	ListNode curNode = lists[cur[1]];
        	lists[cur[1]] = curNode.next;
            curNode.next = null;

        	if (sol == null){
                sol = curNode;
                res = sol;
            } 
            else {
                sol.next = curNode;
                sol = curNode;
            }

	        if (lists[cur[1]] != null){
	        	cur[0] = lists[cur[1]].val;
	        	heap.add(cur);
            }
        }

        return res;
    }
}
